### PR TITLE
[Bug] Change dropdown position of `DatePicker` based on space

### PR DESCRIPTION
**Pull request by nadijagraca**
## Description
- Minor feature app example dataframe fix
- Fix for bug on datepicker dropdown/calendar position raised by Max on: https://github.com/mckinsey/vizro/pull/319#discussion_r1517637922
Solution:
- clientside callback updates the `dropdownPosition` to `top-start` if there is not enough space below the input field to load the datepicker calendar. `dropdownPosition` was set to `bottom-start` as a workaround for issue: [219](https://github.com/snehilvj/dash-mantine-components/issues/219)

## Screenshot
Screenshot 2024-03-13 at 18 33 45

## Notice

- [x] I acknowledge and agree that, by checking this box and clicking "Submit Pull Request":

  - I submit this contribution under the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt) and represent that I am entitled to do so on behalf of myself, my employer, or relevant third parties, as applicable.
  - I certify that (a) this contribution is my original creation and / or (b) to the extent it is not my original creation, I am authorized to submit this contribution on behalf of the original creator(s) or their licensees.
  - I certify that the use of this contribution as authorized by the Apache 2.0 license does not violate the intellectual property rights of anyone else.
  - I have not referenced individuals, products or companies in any commits, directly or indirectly.
  - I have not added data or restricted code in any commits, directly or indirectly.

### CODE REVIEW

File: vizro-core/src/vizro/static/js/models/date_picker.js [hunk 1]
```diff
@@ -7,3 +7,15 @@ export function _update_date_picker_values(value, input_store) {
...
+  var element_id = window.dash_clientside.callback_context.inputs_list[0]["id"];
+  var windowHeight = window.innerHeight;
+
+  var element = document.getElementById(element_id);
+  var rect = element.getBoundingClientRect();
+
+  var position = rect.y + 360 > windowHeight ? "top-start" : "bottom-start";
+
+  return position;
+}
```
Comment (huong-li-nguyen) [inline vizro-core/src/vizro/static/js/models/date_picker.js:21]:
```suggestion
  var element = document.getElementById(element_id);
  var rect = element.getBoundingClientRect();
  var position = rect.y + 360 > window.innerHeight ? "top-start" : "bottom-start";
  return position;
}
```

File: vizro-core/src/vizro/static/js/models/date_picker.js [hunk 2]
```diff
@@ -7,3 +7,15 @@ export function _update_date_picker_values(value, input_store) {
   }
   return [value, value];
 }
+
+export function _update_date_picker_position(clicks) {
+  var element_id = window.dash_clientside.callback_context.inputs_list[0]["id"];
```
Comment (huong-li-nguyen) [inline vizro-core/src/vizro/static/js/models/date_picker.js:12]: How does this inputs_list look like? Like what else is stored there?

File: vizro-core/src/vizro/models/_components/form/date_picker.py [hunk 3]
```diff
@@ -70,6 +70,12 @@ def build(self):
             inputs=inputs,
         )
 
+        clientside_callback(
+            ClientsideFunction(namespace="clientside", function_name="update_date_picker_position"),
+            output=Output(self.id, "dropdownPosition"),
+            inputs=Input(self.id, "n_clicks"),
+        )
+
```
Comment (huong-li-nguyen) [inline vizro-core/src/vizro/models/_components/form/date_picker.py:78]: Could you add a comment above of why we need this work-around? And when we can eventually remove it? Link to issue on the mantine library if it's a bug on their side

**Review by huong-li-nguyen — CHANGES_REQUESTED** [review]
Hey @nadijagraca - thanks for the quick fix!

**1. Could we remove the space between the dropdown and the calendar?**
Screenshot 2024-03-13 at 18 51 56

2. You might have to add another condition if it does not have enough vertical space:
Screenshot 2024-03-13 at 18 50 46

File: vizro-core/src/vizro/static/js/models/date_picker.js (hunk 2 above)
Comment (nadijagraca) [reply vizro-core/src/vizro/static/js/models/date_picker.js:12]: inputs_list contains data about the callback inputs. 

Screenshot 2024-03-14 at 15 18 46

File: vizro-core/src/vizro/models/_components/form/date_picker.py [hunk 4]
```diff
@@ -69,6 +69,13 @@ def build(self):
             output=output,
             inputs=inputs,
         )
+        # clientside callback is used as a workaround for situations where insufficient space below the date-picker
+        # input field prevents the date-picker calendar from opening.
```
Comment (huong-li-nguyen) [inline vizro-core/src/vizro/models/_components/form/date_picker.py:73]:
```suggestion
        # clientside callback is required as a workaround when the date-picker is overflowing its parent container if there
        # is no sufficient space. Caused by another workaround for this issue: https://github.com/snehilvj/dash-mantine-components/issues/219
```

**Pull request merged**